resources/analysisTools/snvPipeline/snv_extractor.pl

```diff
@@ -0,0 +1,217 @@
+#!usr/bin/perl
+
+# By Ivo Buchhalter
+# Hard coded script to extract snvs in the snv pipeline
+# Extracts: somatic snvs -> snvs_PID_somtatic_snvs_conf_?_to_10.vcf
+#           somatic coding snvs -> snvs_PID_somtatic_coding_snvs_conf_?_to_10.vcf
+#           germline coding snvs -> snvs_PID_germline_coding_snvs_conf_?_to_10.vcf
+
+# By Christopher Previti - 20191204
+# Hard coded to extract all (regardless of confidence score) somatic snvs for a list of genes (whitelist)  
+
+
+use strict;
+use warnings;
+use Getopt::Long;
+
+my $infile;
+my $minconf=8;
+my $region = 0;
+my $pid;
+my $extractsyn = 0;
+my $extractNcRNA = 0;
+my $bgzip = "bgzip";
+my $tabix = "tabix";
+my $whitelist = "NA";
+my $whitelist_minconf = 5;
+
+GetOptions (	                "infile=s"	 		=> \$infile,		# vcf file, can be bgzipped
+				"minconf=i"			=> \$minconf,		# minimum confidence score
+				"region=s"			=> \$region,		# file with the region when only a region should be extracted
+				"ncRNA=i"			=> \$extractNcRNA,	# If non coding RNAs (functional) should be extracted
+				"pid=s"				=> \$pid,
+				"synonymous=i"		        => \$extractsyn,
+				"bgzip=s"			=> \$bgzip,
+				"tabix=s"			=> \$tabix,
+                                "whitelist=s"                   => \$whitelist,         # List of whitelisted genes, if defined do extra filtering
+                                "whitelist_minconf=i"           => \$whitelist_minconf  # Minimum confidence score for variants in whitelisted genes, is only used if a whitelist has been specified
+ 
+) or die "Could not get the options!\n";
+
+if($region ne "0" && (!-f $region || $infile !~ /\.gz$/ || !-f $infile.".tbi")){die "region-file: $region is not a valid file or, infile: $infile is not zipped or there is no index for the infile $infile.tbi\n";}
+if(!-f $infile){die "The provided infile: $infile is not a valid file\n";}
+if(-f $infile && $infile =~ /\.gz$/) { open(IN, "zcat $infile |") or die "Could not open the infile: $infile\n"; }
+else{open(IN, "<$infile") or die "Could not open the infile: $infile\n";}
+
+if($region ne "0"){print "Region file provided: $region\n";}
+
+my $outsom = $pid."_somatic_snvs_conf_".$minconf."_to_10.vcf";
+my $outsomcod = $pid."_somatic_functional_snvs_conf_".$minconf."_to_10.vcf";
+my $outgermcod = $pid."_germline_functional_snvs_conf_".$minconf."_to_10.vcf";
+my $outsyn = $pid."_somatic_functional_and_synonymous_snvs_conf_".$minconf."_to_10.vcf";
+my $outNcRNA = $pid."_somatic_functional_ncRNA_snvs_conf_".$minconf."_to_10.vcf";
+
+my $whitelist_maxconf=$minconf-1;
+
+
+open(SOM, ">$outsom") or die "Could not open the file $outsom\n";
+open(COD, ">$outsomcod") or die "Could not open the file $outsomcod\n";
+open(GER, ">$outgermcod") or die "Could not open the file $outgermcod\n";
+
+if($extractsyn == 1){open(SYN, ">$outsyn") or die "Could not open the file $outsyn\n";}
+if($extractNcRNA == 1){open(NCRNA, ">$outNcRNA") or die "Could not open the file $outNcRNA\n";}
+
+my $head;
+while(<IN>)
```

Comment:
See other comment for safer Perl IO.

Reply:
See [here](https://gitlab.com/one-touch-pipeline/otp/-/wikis/Workflow-Development/Coding-Guidelines#perl) how to do safe IO in Perl. The `<>` operator is evil, because it returns `undef` both if the file is ends (i.e. `eof(IN) == true`) and if there is an IO error.

Please avoid the implicit or explicit usage of the "$_" variable. This is short, but at the cost of readability.